include/Probe.php

```diff
@@ -0,0 +1,1114 @@
+		$xrd = parse_xml_string($data, false);
+
+		if (!is_object($xrd)) {
+			// If it is not XML, maybe it is JSON
+			$webfinger = json_decode($data, true);
+
+			if (!isset($webfinger["links"]))
+				return false;
+
+			return $webfinger;
+		}
+
+		$xrd_arr = xml::element_to_array($xrd);
+		if (!isset($xrd_arr["xrd"]["link"]))
+			return false;
+
+		$webfinger = array();
+
+		if (isset($xrd_arr["xrd"]["subject"]))
+			$webfinger["subject"] = $xrd_arr["xrd"]["subject"];
+
+		if (isset($xrd_arr["xrd"]["alias"]))
+			$webfinger["aliases"] = $xrd_arr["xrd"]["alias"];
+
+		$webfinger["links"] = array();
+
+		foreach ($xrd_arr["xrd"]["link"] AS $value => $data) {
+			if (isset($data["@attributes"]))
+				$attributes = $data["@attributes"];
+			elseif ($value == "@attributes")
+				$attributes = $data;
+			else
+				continue;
+
+			$webfinger["links"][] = $attributes;
+		}
+		return $webfinger;
+	}
+
+	/**
+	 * @brief Poll the noscrape page (Friendica specific)
+	 *
+	 * @param string $noscrape Link to the noscrape page
+	 * @param array $data The already fetched data
+	 *
+	 * @return array noscrape data
+	 */
+	private function poll_noscrape($noscrape, $data) {
+		$content = fetch_url($noscrape);
+		if (!$content)
+			return false;
+
+		$json = json_decode($content, true);
+		if (!is_array($json))
+			return false;
+
+		if (isset($json["fn"]))
+			$data["name"] = $json["fn"];
+
+		if (isset($json["addr"]))
+			$data["addr"] = $json["addr"];
+
+		if (isset($json["nick"]))
+			$data["nick"] = $json["nick"];
+
+		if (isset($json["comm"]))
+			$data["community"] = $json["comm"];
+
+		if (isset($json["tags"])) {
+			$keywords = implode(" ", $json["tags"]);
+			if ($keywords != "")
+				$data["keywords"] = $keywords;
+		}
+
+		$location = formatted_location($json);
+		if ($location)
+			$data["location"] = $location;
+
+		if (isset($json["about"]))
+			$data["about"] = $json["about"];
+
+		if (isset($json["key"]))
+			$data["pubkey"] = $json["key"];
+
+		if (isset($json["photo"]))
+			$data["photo"] = $json["photo"];
+
+		if (isset($json["dfrn-request"]))
+			$data["request"] = $json["dfrn-request"];
+
+		if (isset($json["dfrn-confirm"]))
+			$data["confirm"] = $json["dfrn-confirm"];
+
+		if (isset($json["dfrn-notify"]))
+			$data["notify"] = $json["dfrn-notify"];
+
+		if (isset($json["dfrn-poll"]))
+			$data["poll"] = $json["dfrn-poll"];
+
+		return $data;
+	}
+
+	/**
+	 * @brief Check for valid DFRN data
+	 *
+	 * @param array $data DFRN data
+	 *
+	 * @return int Number of errors
+	 */
+	public static function valid_dfrn($data) {
+		$errors = 0;
+		if(!isset($data['key']))
+			$errors ++;
+		if(!isset($data['dfrn-request']))
+			$errors ++;
+		if(!isset($data['dfrn-confirm']))
+			$errors ++;
+		if(!isset($data['dfrn-notify']))
+			$errors ++;
+		if(!isset($data['dfrn-poll']))
+			$errors ++;
+		return $errors;
+	}
+
+	/**
+	 * @brief Fetch data from a DFRN profile page
+	 *
```

Comment:
Some more description would be nice: We try to get the profile information via noscrape (json request) if this is not possible we will try the scrape the profile data directly form the vcard widget.